### ISSUE TITLE
Unify configuration logic with package:jnigen through package:config

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/ffigen#528]
Our current configuration provider is config-file only (with a single command-line override for compiler options).

To unify the experience between different packages, we would like to unify the config providing logic.

We could do this by introducing a shared `package:config` that is opinionated about (1) the layout of config files (JSON/YAML, snake_cased or dash-cased identifiers), (2) overrides through cli `-Dsome.key=value` and environment variables `SOME__KEY=value`.

I've created a prototype for this on https://github.com/dcharkes/config

Applying this logic in our https://github.com/dart-lang/ffigen/tree/master/lib/src/config_provider, requires some massaging of logic.

### Not loadable as dynamic

Currently the validation of values assumes they can be loaded as `dynamic` and be cast later.

This will not work for the following reasons:

1. For environment variables / CLI arguments we might need to pass a separator token for splitting if we want `List<String>` or `String`.
2. We need to define for every lookup whether we join the defines from all three sources (for example include paths would be joined, while some boolean flag from the command line would override the boolean flag in the config file.

We will need to rewrite the `extractor`/`validator` in the spec to provide the right information. For example `config.getStringList(key, cliSeparator: ';', combineAll: true)`.

### Validation

Currently, all types of validation happen in the config provider. This contains two types of validation:

1. Whether something is of the right type, for example whether it's a boolean.
2. FFIgen specific validation, for example `dartClassNameValidator`.

The first type of validation logic can be done by `package:config`, while the second type needs to be in `package:ffigen`.

I haven't made a plan yet on what is the right type of abstraction to this validation.

One thing we could think about is we want the error reporting to be done by `package:config` in a consistent way, that we would provide some kind of callbacks API. That way we can make sure we always use the same way to represent the paths, the same sentence structure etc.

### Extraction

Currently, `package:config` only supports overriding config from CLI/environment for String/List<String>/bool/int. The rest must be provided by a config file. However, even when it's provided by the config file, it would still be transformed by `package:ffigen` to a data structure (which could potentially fail).

I'm thinking that validate and extract should be combined into a single function that we then provide as a callback. Then it could be also type-safe.
For these to be combined we need to either use exceptions to communicate validation failures (which would then be caught and outputted to the logger in `package:ffigen`) or we always return nullable values add an `void Function(String) errorReporter` function to the config.

### Related

* JNIgen with `package:config`
    * https://github.com/dart-lang/jnigen/pull/197
* JNIgen open issue to have better validation
    * https://github.com/dart-lang/jnigen/issues/186

**Comment by dcharkes** [comment]
> I'm thinking that validate and extract should be combined into a single function that we then provide as a callback. Then it could be also type-safe.
> For these to be combined we need to either use exceptions to communicate validation failures (which would then be caught and outputted to the logger in `package:ffigen`) or we always return nullable values add an `void Function(String) errorReporter` function to the config.

Having a callback that sets a late field as we do in FFIgen works fairly well if you want to combine `package:config` checks and manual checks. The initialization logic can then just catch all `FormatExceptions` for all callbacks and combine them in the end. That way we get all error messages instead of stopping errors after the first exception.

The callback can refer to previous late fields set (but will throw a late initialization error if loading that field failed.) For example if some field is required if another field has a certain value.

I've built one `Config`-wrapper here: https://github.com/dcharkes/native_assets_cli/blob/a4e4fad0c96a4fe8b0e2b8509e84ca56efbe3a97/lib/src/model/native_assets_cli_config.dart

The callbacks are fairly similar to the `Specification`s in FFIgen. The main difference them not splitting the steps in different parts.

To deal with checking `Config`-wrappers having other constructors that initialize all the late fields, I propose roundtripping via JSON and a `Config` object in a factory. That way always get consistent error messages.
(The alternative would be introducing a second callback per field. The first callback would extract a type `T` out of the `Config` and assign to the late field. The second would read the late field and throw more errors. The downsides of this approach is that the field names have to be typed more than once, the closure syntax has to be typed twice, and the config key has to be manually added to an error message for context.)

**Comment by dcharkes** [comment]
https://github.com/dart-lang/tools/pull/45

@mannprerak2 should we consider doing a breaking change to make our YAML identifiers use underscores instead of dashes? (We can still keep the dashes working as a fallback, but give a warning on it.) It seems that using underscores is more common in the YAML config files I've seen. Most likely because it's compatible with environment variables and commandline args.

Also, most syntaxes use `.`s for navigating hierarchy. We use `>` in our documentation. This would just be a documentation change. But if we adopt `package:cli_config` here, the documentation should be consistent with the CLI overrides. 😄

**Comment by mannprerak2** [comment]
> should we consider doing a breaking change to make our YAML identifiers use underscores instead of dashes?

Yeah I guess we can, but imo handling underscores and hyphen is probably better, it seems like a hassle for the end user.

**Comment by TzviPM** [comment]
I was just about to suggest exactly this idea. having a unified way to configure these tools both by yaml and cli args with a clear mapping between them will make it much easier to build tooling around them that has a non-yaml way of doing configuration
